### DISCUSSION TITLE
[Android/Wasm] Laundry list - CalendarView, Pivot, NavigationView, ListView

### DISCUSSION THREAD

**Discussion by emceelovin** [unoplatform/uno#8673] [General]
NavigationView hamburger button overlaps it's content area on wasm, android, and windows when the screen size is small. Which you can see in the images below. Also, the pane opens by default unless you set IsPaneOpen="False". The hamburger icon on android is a random glyph as well.

CalendarView acts absolutely horrible when stretched both horizontally and vertically inside of both a Page and Grid as the only item. Extreme lag, random dates, random sizing, etc. On Windows it's totally fine. <CalendarView HorizontalAlignment="Stretch" VerticalAlignment="Stretch"/> spazzes out on Android/WASM. <CalendarView/> is fine though.
![278838471_1399422367167832_1139040643768594256_n](https://user-images.githubusercontent.com/55720292/166336244-60cec0f5-3148-449f-8f34-eed464417be1.jpg)
![image](https://user-images.githubusercontent.com/55720292/166336260-3eb7ae75-9f72-4e79-ae12-a4f49ba54b27.png)

Pivot doesn't size PivotItem content to the width of the screen on Android. It sizes the PivotItem to span the width of the overflowed Pivot control itself. As you can see below, the PivotItem on mobile/Android spans all the way to the width of the headers section:
[Android]
![278980129_491863469341008_203870854701216255_n](https://user-images.githubusercontent.com/55720292/166334871-4a6548e3-2391-4d12-9990-9c98687a6e99.jpg)
![278972943_692435358674685_2600733038223996257_n](https://user-images.githubusercontent.com/55720292/166334885-4dc84ad5-d9e9-4fa5-bcbb-5a8e3081d9a0.jpg)

[Windows]
![image](https://user-images.githubusercontent.com/55720292/166335002-974da941-d6f3-431b-ac4b-7c1873801b7f.png)

[WASM]
![image](https://user-images.githubusercontent.com/55720292/166336537-ea8d3d6a-8c96-4c5e-881e-494fbb73c367.png)

Code for the laundry list is based off a fresh blank template:

MainPage.xaml.cs
```xaml
<Page
    x:Class="OneGreatTeam.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:OneGreatTeam"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d"    
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
    <NavigationView IsPaneOpen="False">
        <Pivot>
            <PivotItem Header="ListView">
                <Page>
                    <ListView
                        SelectionMode="Single"
                        ItemsSource="{x:Bind ViewModel.Items}">
                        <ListView.ItemTemplate>
                            <DataTemplate x:DataType="local:ListViewItem">
                                <Grid RowDefinitions="auto" ColumnDefinitions="*,*">
                                    <TextBlock Foreground="White" Text="{Binding FirstName}" Grid.Row="0" Grid.Column="0" HorizontalAlignment="Center"/>
                                    <TextBlock Foreground="White" Text="{Binding LastName}" Grid.Row="0" Grid.Column="1" HorizontalAlignment="Center"/>
                                </Grid>
                            </DataTemplate>
                        </ListView.ItemTemplate>
                    </ListView>
                </Page>
            </PivotItem>
            <PivotItem Header="CalendarView">
                <CalendarView HorizontalAlignment="Stretch" VerticalAlignment="Stretch"/>
            </PivotItem>
            <PivotItem Header="Header 3">
                <Page Background="Blue"/>
            </PivotItem>
            <PivotItem Header="Header 4">
                <Page Background="Yellow"/>
            </PivotItem>
            <PivotItem Header="Header 5">
                <Page Background="Pink"/>
            </PivotItem>
        </Pivot>
    </NavigationView>
</Page>
```

Other than these issues, HUGE improvement noticed in Uno with the last couple releases are so. You guys are doing an absolutely amazing job. Keep up the good work.

**Comment by jeromelaban** [comment]
Thanks for the feedback :) Note that you're using `ColumnDefinitions="*,*"` which is only supported in the upcoming 4.3. You can try it with the 4.3-dev bits. For the rest, those would be individual issues that you can open for us to track more easily.

**Reply by emceelovin** [reply]
Ahhh, I wasn't aware of the issues with shorthand stuff like that. Is there any page that has a WinUI vs Uno WinUI differences? Honestly been in the MAUI world a lot lately. I will change that and see what happens! Any timeframe guess on when 4.3 will be released? Also, how is the AOT coming along as far as it working by default without having to do anything? I haven't been able to get it working still. I'm going to add another issue, probably to this one, once I have some time. Frame navigation is extremely sluggish, but it might be due to the controls having issues themselves. Will report back!

**Reply by emceelovin** [reply]
Swithing the column/row definition stuff fixed that issue. Thanks! Once that PivotItem issue is fixed, that'll really make things wonderful! However, there are still some rendering issues(so far just on android). Will report back later with more issues etc.

If I want to try the latest stuff from 4.3.x, which packages do I have to update to 4.3-dev btw?

**Reply by jeromelaban** [reply]
> Also, how is the AOT coming along as far as it working by default without having to do anything?

Which platform are you referring to?


> If I want to try the latest stuff from 4.3.x, which packages do I have to update to 4.3-dev btw?

Any latest 4.3.0-dev.XX package will do!

> Frame navigation is extremely sluggish,

What do you mean by sluggish? On which platform? Can you make a video of this?

**Reply by emceelovin** [reply]
There is a pretty big delay when navigating to pages on WASM, and I think android(will report back). I will make a video of anything I can, sure! navigating to pages via Pivot is great, so it seems like it's just a code issue possibly. 

so just updating to any package will work, awesome. 

about the AOT, I meant the WASM AOT. sorry about that.

**Reply by jeromelaban** [reply]
I see. For Wasm, depending on the complexity of the target page, the navigation may take some time if the build is running with the interpreter. You can definitely use AOT to get better performance.

**Reply by emceelovin** [reply]
that's what I figured, and noticed. the more complex the page, the slower it gets. I still can't seem to get AOT working with wasm for some reason. I'll try again later on. I wanna say it sits there doing something, and then fails to generate the AOT layout. I'll see if I can get more errors. I followed the old and new docs, but it doesn't seem to work. I'd like to get everything setup and running as fast as possible using everything. that way I can tell if it's an actual bug, or just my environment setup.

**Reply by jeromelaban** [reply]
Thanks. Let us know which error you're getting, could be docs related.

**Reply by emceelovin** [reply]
Will do. I remember before, I was using a doc that was outdated and had stuff not needed. Anyway you can link me to an updated doc for the latest 4.2 release? Possibly with the steps from a fresh new project. I started with a fresh new template as well as having the latest templates. Not really looking for my hand-held. I just can't seem to find the right docs.

**Reply by emceelovin** [reply]
I might have got it to work. A bunch of clang++'s popped up and are building something it seems. it's taking a while, but hopefully i'll finally have AOT for WASM and be on the same page as everyone else. It mentions it takes a while, but for a blank project, decent machine, how long is "a while?"

[EDIT: it looks like when I copied the one project setting off the docs it didnt copy. I have the same thing twice]
All of the clang+++'s are done, but now wasm-Id is using up a TON of memory and hasn't gone away. This is my project file:
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <NoWarn>NU1701</NoWarn>
	  <WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)'=='Debug'">
    <MonoRuntimeDebuggerEnabled>true</MonoRuntimeDebuggerEnabled>
    <DefineConstants>$(DefineConstants);TRACE;DEBUG</DefineConstants>
    <DebugType>portable</DebugType>
    <DebugSymbols>true</DebugSymbols>
    <!--
		IL Linking is disabled in Debug configuration.
		When building in Release, see https://platform.uno/docs/articles/features/using-il-linker-webassembly.html
		-->
    <WasmShellILLinkerEnabled>false</WasmShellILLinkerEnabled>
	<WasmShellMonoRuntimeExecutionMode Condition="'$(Configuration)'=='Release'">InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
  </PropertyGroup>
  <ItemGroup>
    <Content Include="Assets\SplashScreen.png" />
  </ItemGroup>
  <ItemGroup>
    <UpToDateCheckInput Include="..\OneGreatTeam.Shared\**\*.xaml" />
  </ItemGroup>
  <ItemGroup>
    <EmbeddedResource Include="WasmCSS\Fonts.css" />
    <EmbeddedResource Include="WasmScripts\AppManifest.js" />
  </ItemGroup>
  <ItemGroup>
    <LinkerDescriptor Include="LinkerConfig.xml" />
  </ItemGroup>
  <ItemGroup>
    <!--
		This item group is required by the project template because of the
		new SDK-Style project, otherwise some files are not added automatically.

		You can safely remove this ItemGroup completely.
		-->
    <None Include="Program.cs" />
    <None Include="LinkerConfig.xml" />
    <None Include="wwwroot\web.config" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Windows.Compatibility" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging" Version="5.0.0" />
    <PackageReference Include="Uno.Extensions.Logging.WebAssembly.Console" Version="1.3.0" />
    <PackageReference Include="Uno.WinUI.WebAssembly" Version="4.2.6" />
    <PackageReference Include="Uno.WinUI.RemoteControl" Version="4.2.6" Condition="'$(Configuration)'=='Debug'" />
    <PackageReference Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="4.2.6" />
    <PackageReference Include="Uno.Wasm.Bootstrap" Version="3.3.1" />
    <PackageReference Include="Uno.Wasm.Bootstrap.DevServer" Version="3.3.1" />
  </ItemGroup>
  <Import Project="..\OneGreatTeam.Shared\OneGreatTeam.Shared.projitems" Label="Shared" Condition="Exists('..\OneGreatTeam.Shared\OneGreatTeam.Shared.projitems')" />
</Project>
```

**Reply by emceelovin** [reply]
It's stuck at this process. If i end wasm-ld.exe, the server and app will launch. I'm not sure how to debug this hanging process to offer any help unfortunately. I updated VS to Preview 6 yesterday. Do you think that has anything to do with anything?
![image](https://user-images.githubusercontent.com/55720292/167217473-412bf6d8-cd2f-460f-a94e-18b4ba2c59b7.png)

**Reply by jeromelaban** [reply]
You're hitting one of the wasm linker issue, which can happen when there's a lot of code to link. You can adjust the wasm linker optimization level using this: https://github.com/unoplatform/Uno.Wasm.Bootstrap#emscripten-linker-optimizations-flags.

The default setting (full linking) generates the most optimized code, but also is the slowest to run.

**Reply by emceelovin** [reply]
Oh that's the linker. Gotcha. I'll give that a try. It literally took 5-10 minutes to link.

**Reply by emceelovin** [reply]
I added the project configuration setting to completely disable IL linking, as well as any linking optimization, and wasm-ld still runs forever(endless) until i end it, and then the browser will popup. From what was actually optimized, half AOT is INSANELY fast. hats off. Looking forward to the future where everything just works. Any tips for how to disable the IL linking period, or do things have to be linked together like a regular C/C++ binary etc? I don't know how things operate in the WASM world.

**Comment by emceelovin** [comment]
@jeromelaban any update on the Pivot/CalendarView issue, or anyway I can track it somehow? Just curious if it's being worked on etc.

**Comment by emceelovin** [comment]
bump

@jeromelaban I haven't really spent much time on the UI end of things lately, and have had a lot going on in life, but I've updated to the latest Uno, and am even more impressed than I have been. Great work as always, but I'm still experiencing two "bugs" so far. I'm including a video to show them. Page's inside of a PivotItem aren't sized to fit the width of the screen properly when the PivotItem's takeup more than the width, and the NavigationView hamburger open/close button overlaps the content of a NavigationView(Pivot in this case). Let me know if you need anything else to help!

https://user-images.githubusercontent.com/55720292/202865277-6adea3f9-7dcc-4798-833a-8f66f4de289f.mp4

**Reply by jeromelaban** [reply]
Thanks for the report! We've made a lot of progress for sure, and there's plenty left to do!

Could you share the repro you used to create this video, we'll open an issue with it.

**Reply by emceelovin** [reply]
Sure thing! If it's my XAML, then I'm totally ok with it, but this is a fresh new "kitchen sink" test project to see the current progress. The same behavior happens on Android as well. When you swipe to make the Pivot control Pivot like it did on Windows Phone, it doesn't pivot. The whole control just scrolls horizontally as if its in a ScrollView.

[UnoApp.zip](https://github.com/unoplatform/uno/files/10053640/UnoApp.zip)

**Reply by jeromelaban** [reply]
Thanks for the report. You'll notice that the Windows version of the app does the same thing:
![image](https://user-images.githubusercontent.com/5839577/203802344-7c28136a-4f49-4549-8c45-ec0e03d92089.png)

Which would tend to indicate more of a problem in the way the navigation view content is handling size. My guess would be that you should be using an `AdaptiveTrigger` to include a padding at the top, above the Pivot control.

**Reply by emceelovin** [reply]
Yeah, the windows application does the same. As they both have since the original post. I just expected NavigationView to handle content how it handles it in WinUI and MAUI does with Shell, so I'm not sure if it was me missing something, a bug, etc. I did think of what you mentioned and was gonna do something based off the size, but I can probably workaround that. The Pivot issue not sizing children properly - which makes it not even allow Pivot cycling - is another issue. Any news on that?

**Reply by jeromelaban** [reply]
We have not worked on Pivot changes at this time. It's probably best if you open an issue, though this looks very similar to https://github.com/unoplatform/uno/issues/4866.
